src/python/foglamp/core/browser.py

```diff
@@ -0,0 +1,227 @@
+  http://<address>/foglamp/asset/{asset}/{reading}/summary
+    - Return a summary (min, max and average) for the specified asset and sensor
+
+  All but the /foglamp/asset API call take a set of optional query parameters
+    limit=x     Return the first x rows only
+    seconds=x   Limit the data return to be less than x seconds old
+    minutes=x   Limit the data returned to be less than x minutes old
+    hours=x     Limit the data returned to be less than x hours old
+
+  Note seconds, minutes and hours can not be combined in a URL. If they are then only seconds
+  will have an effect.
+
+  TODO: Improve error handling, use a connection pool
+"""
+
+import asyncpg
+import asyncio
+import json
+from aiohttp import web
+
+__author__ = "Mark Riddoch"
+__copyright__ = "Copyright (c) 2017 OSIsoft, LLC"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+__DB_URL = 'postgresql://foglamp:foglamp@localhost:5432/foglamp'
+__DEFAULT_LIMIT = 20
+__TIMESTAMP_FMT = 'YYYY-MM-DD HH24:MI:SS.MS'
+
+def setup(app):
+    """
+    Add the routes for the API endpoints supported by the data browser
+    """
+    app.router.add_route('GET', '/foglamp/asset', asset_counts)
+    app.router.add_route('GET', '/foglamp/asset/{asset_code}', asset)
+    app.router.add_route('GET', '/foglamp/asset/{asset_code}/{reading}', asset_reading)
+    app.router.add_route('GET', '/foglamp/asset/{asset_code}/{reading}/summary', asset_summary)
+
+
+async def asset_counts(request):
+    """
+    Browse all the asserts for which we have recorded readings and
+    return a readings count.
+
+    Return the result of the Postgres query 
+    select asset_code, count from readings group by asset_code;
+    """
+
+    conn = await asyncpg.connect(__DB_URL)
+
+    # Select the assets from the readings table
+    rows = await conn.fetch(
+        'select asset_code, count(*) from readings group by asset_code')
+    columns = ('asset_code', 'count')
+    results = []
+    for row in rows:
+      results.append(dict(zip(columns, row)))
+
+    # Close the connection.
+    await conn.close()
+
+    return web.json_response(results);
+
+async def asset(request):
+    """
+    Browse a particular assert for which we have recorded readings and
+    return a readings with timestamps for the asset. The number of readings
+    return is defaulted to a small number (20), this may be changed by supplying
+    the query parameter ?limit=xx
+
+    Return the result of the Postgres query 
+    select timestamp, reading from readings where asset_code = 'asset_code' order by user_ts limit 20
+    """
+
+    conn = await asyncpg.connect(__DB_URL)
+    asset_code =  request.match_info.get('asset_code', '')
+    limit = __DEFAULT_LIMIT
+    if 'limit' in request.query:
+      limit = request.query['limit']
+
+    query = 'select to_char(user_ts, \'{0}\') as "timestamp", (reading)::json from readings where asset_code = \'{1}\''.format(__TIMESTAMP_FMT, asset_code)
```

Comment:
and actually, use ...  asyncpg uses the native PostgreSQL syntax for query arguments: $n. e.g. `'SELECT * FROM users WHERE name = $1', 'Bob')`